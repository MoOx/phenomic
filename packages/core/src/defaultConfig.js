import path from "path";

export default {
  path: process.cwd(),
  outdir: path.join(process.cwd(), "dist"),
  port: 3333,
  bundleName: "phenomic.main",
  plugins: []
};
