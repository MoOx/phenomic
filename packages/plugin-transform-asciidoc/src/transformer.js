import asciidoctor from "asciidoctor.js";
import unified from "unified";
import parse from "rehype-parse";
import slug from "rehype-slug";
import autoLinkHeadings from "rehype-autolink-headings";
import highlight from "rehype-highlight";
import html from "rehype-stringify";
import rehype2react from "rehype-react";
import deburr from "lodash.deburr";
import kebabCase from "lodash.kebabcase";

const defaultOptions = {
  safe: "secure",
  attributes: {
    header_footer: "false",
    "linkcss!": "",
    "copycss!": ""
  }
};

const envAttributes = [
  "asciidoctor",
  "asciidoctor-version",
  "backend",
  "basebackend",
  "docdate",
  "docdatetime",
  "docdir",
  "docfile",
  "docfilesuffix",
  "docname",
  "doctime",
  "doctype",
  "docyear",
  "embedded",
  "filetype",
  "htmlsyntax",
  "localdate",
  "localdatetime",
  "localtime",
  "localyear",
  "outdir",
  "outfile",
  "outfilesuffix",
  "safe-mode-level",
  "safe-mode-name",
  "safe-mode-unsafe",
  "safe-mode-safe",
  "safe-mode-server",
  "safe-mode-secure",
  "user-home"
];

function processAsciidoc(text) {
  const ad = asciidoctor();
  const doc = ad.load(text, defaultOptions);
  let data = {
    ...doc.attributes.$$smap,
    contents: ad.convert(text, defaultOptions)
  };

  data.date =
    doc.getAttribute("date") || doc.getAttribute("revdate") || undefined;
  data.title = doc.getAttribute("doctitle");
  data.layout = doc.getAttribute("layout");
  data.showdate = doc.getAttribute("nodate", true);
  data.tags = doc.getAttribute("tags");
  data.tags = data.tags
    ? data.tags.split(",").map(tag => kebabCase(deburr(tag)))
    : [];

  envAttributes.map(key => delete data[key]);

  return data;
}

// here we optimize structure just a little to have to smallest json possible
const createElement = (component, props, children) => {
  return {
    t: component,
    ...(props && Object.keys(props).length ? { p: props } : {}),
    ...(children ? { c: children } : {})
  };
};

export default (config?: PhenomicConfig, contents: string) => {
  const data = processAsciidoc(contents);

  const processor = unified()
    .use(parse, { fragment: true })
    .use(slug)
    .use(autoLinkHeadings)
    .use(highlight);

  const useReact =
    config &&
    config.plugins &&
    config.plugins.find(p => p.name === "@phenomic/plugin-renderer-react");

  if (!useReact) {
    processor.use(html);
  } else {
    processor.use(rehype2react, { createElement });
  }

  data.contents = processor.processSync(data.contents).contents;
  return data;
};
