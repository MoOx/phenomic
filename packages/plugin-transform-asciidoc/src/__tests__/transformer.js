import defaultConfig from "@phenomic/core/lib/defaultConfig.js";

import transformer from "../transformer.js";

const input = `
= Hello, AsciiDoc!
Ezra Lalonde <ezra@usefuliftrue.com>
v1.0, 2017-10-05: First Draft
:tags: phenomic, documentation, test, déjà vu
:iconsdir: /path/to/icons

Test of Asciidoc.js transformer for Phenomic.

== Heading

* item 1
* item 2

[source,javascript]
const name = "Ezra"
console.log(\`Hello, ${name}!\`)
`;

it("should render asciidoc as json (react component)", () => {
  const adoc = transformer(
    {
      ...defaultConfig,
      plugins: [{ name: "@phenomic/plugin-renderer-react" }]
    },
    input
  ).contents;
  expect(adoc).toMatchSnapshot();
});

it("should render asciidoc as html", () => {
  const adoc = transformer(defaultConfig, input);
  expect(adoc).toMatchSnapshot();
});

it("should have tags", () => {
  const adoc = transformer(
    defaultConfig,
    `
= Title
:tags: phenomic, documentation, test, déjà vu
  `
  );
  expect(adoc.tags).toEqual(["phenomic", "documentation", "test", "deja-vu"]);
});

it("should not have tags", () => {
  const adoc = transformer(
    defaultConfig,
    `
= Title
  `
  );
  expect(adoc.tags).toEqual([]);
});

it("should use explicit date attribute for date", () => {
  const adoc = transformer(
    defaultConfig,
    `
= Title
Ezra Lalonde <ezra@usefuliftrue.com>
v1.0, 2017-10-05: First Draft
:date: 2011-12-25
  `
  );
  expect(adoc.date).toBe("2011-12-25");
});

it("should use revdate for date", () => {
  const adoc = transformer(
    defaultConfig,
    `
= Title
Ezra Lalonde <ezra@usefuliftrue.com>
v1.0, 2017-10-05: First Draft
  `
  );
  expect(adoc.date).toBe("2017-10-05");
});

it("should use have no date attribute", () => {
  const adoc = transformer(
    defaultConfig,
    `
= Title
  `
  );
  expect(adoc.date).toBe(undefined);
});
