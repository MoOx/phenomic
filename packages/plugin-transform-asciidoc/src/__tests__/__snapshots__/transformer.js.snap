// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render asciidoc as html 1`] = `
Object {
  "appendix-caption": "Appendix",
  "appendix-refsig": "Appendix",
  "attribute-missing": "skip",
  "attribute-undefined": "drop-line",
  "author": "Ezra Lalonde",
  "authorcount": 1,
  "authorinitials": "EL",
  "authors": "Ezra Lalonde",
  "backend-html5": "",
  "backend-html5-doctype-article": "",
  "basebackend-html": "",
  "basebackend-html-doctype-article": "",
  "caution-caption": "Caution",
  "chapter-refsig": "Chapter",
  "contents": "<div id=\\"preamble\\">
<div class=\\"sectionbody\\">
<div class=\\"paragraph\\">
<p>Test of Asciidoc.js transformer for Phenomic.</p>
</div>
</div>
</div>
<div class=\\"sect1\\">
<h2 id=\\"_heading\\"><a aria-hidden=\\"true\\" href=\\"#_heading\\"><span class=\\"icon icon-link\\"></span></a>Heading</h2>
<div class=\\"sectionbody\\">
<div class=\\"ulist\\">
<ul>
<li>
<p>item 1</p>
</li>
<li>
<p>item 2</p>
</li>
</ul>
</div>
<div class=\\"listingblock\\">
<div class=\\"content\\">
<pre class=\\"highlight\\"><code class=\\"hljs language-javascript\\" data-lang=\\"javascript\\"><span class=\\"hljs-keyword\\">const</span> name = <span class=\\"hljs-string\\">\\"Ezra\\"</span>
<span class=\\"hljs-built_in\\">console</span>.log(<span class=\\"hljs-string\\">\`Hello, nodejs!\`</span>)</code></pre>
</div>
</div>
</div>
</div>",
  "date": "2017-10-05",
  "doctitle": "Hello, AsciiDoc!",
  "doctype-article": "",
  "email": "ezra@usefuliftrue.com",
  "example-caption": "Example",
  "figure-caption": "Figure",
  "filetype-html": "",
  "firstname": "Ezra",
  "header_footer": "false",
  "iconfont-remote": "",
  "iconsdir": "/path/to/icons",
  "important-caption": "Important",
  "last-update-label": "Last updated",
  "lastname": "Lalonde",
  "layout": undefined,
  "manname-title": "NAME",
  "max-attribute-value-size": 4096,
  "max-include-depth": 64,
  "note-caption": "Note",
  "notitle": "",
  "prewrap": "",
  "revdate": "2017-10-05",
  "revnumber": "1.0",
  "revremark": "First Draft",
  "sectids": "",
  "section-refsig": "Section",
  "showdate": true,
  "stylesdir": ".",
  "stylesheet": "",
  "table-caption": "Table",
  "tags": Array [
    "phenomic",
    "documentation",
    "test",
    "deja-vu",
  ],
  "tip-caption": "Tip",
  "title": "Hello, AsciiDoc!",
  "toc-placement": "auto",
  "toc-title": "Table of Contents",
  "untitled-label": "Untitled",
  "version-label": "Version",
  "warning-caption": "Warning",
  "webfonts": "",
}
`;

exports[`should render asciidoc as json (react component) 1`] = `
Object {
  "c": Array [
    Object {
      "c": Array [
        "
",
        Object {
          "c": Array [
            "
",
            Object {
              "c": Array [
                "
",
                Object {
                  "c": Array [
                    "Test of Asciidoc.js transformer for Phenomic.",
                  ],
                  "t": "p",
                },
                "
",
              ],
              "p": Object {
                "className": "paragraph",
              },
              "t": "div",
            },
            "
",
          ],
          "p": Object {
            "className": "sectionbody",
          },
          "t": "div",
        },
        "
",
      ],
      "p": Object {
        "id": "preamble",
      },
      "t": "div",
    },
    "
",
    Object {
      "c": Array [
        "
",
        Object {
          "c": Array [
            Object {
              "c": Array [
                Object {
                  "p": Object {
                    "className": "icon icon-link",
                  },
                  "t": "span",
                },
              ],
              "p": Object {
                "aria-hidden": true,
                "href": "#_heading",
              },
              "t": "a",
            },
            "Heading",
          ],
          "p": Object {
            "id": "_heading",
          },
          "t": "h2",
        },
        "
",
        Object {
          "c": Array [
            "
",
            Object {
              "c": Array [
                "
",
                Object {
                  "c": Array [
                    "
",
                    Object {
                      "c": Array [
                        "
",
                        Object {
                          "c": Array [
                            "item 1",
                          ],
                          "t": "p",
                        },
                        "
",
                      ],
                      "t": "li",
                    },
                    "
",
                    Object {
                      "c": Array [
                        "
",
                        Object {
                          "c": Array [
                            "item 2",
                          ],
                          "t": "p",
                        },
                        "
",
                      ],
                      "t": "li",
                    },
                    "
",
                  ],
                  "t": "ul",
                },
                "
",
              ],
              "p": Object {
                "className": "ulist",
              },
              "t": "div",
            },
            "
",
            Object {
              "c": Array [
                "
",
                Object {
                  "c": Array [
                    "
",
                    Object {
                      "c": Array [
                        Object {
                          "c": Array [
                            Object {
                              "c": Array [
                                "const",
                              ],
                              "p": Object {
                                "className": "hljs-keyword",
                              },
                              "t": "span",
                            },
                            " name = ",
                            Object {
                              "c": Array [
                                "\\"Ezra\\"",
                              ],
                              "p": Object {
                                "className": "hljs-string",
                              },
                              "t": "span",
                            },
                            "
",
                            Object {
                              "c": Array [
                                "console",
                              ],
                              "p": Object {
                                "className": "hljs-built_in",
                              },
                              "t": "span",
                            },
                            ".log(",
                            Object {
                              "c": Array [
                                "\`Hello, nodejs!\`",
                              ],
                              "p": Object {
                                "className": "hljs-string",
                              },
                              "t": "span",
                            },
                            ")",
                          ],
                          "p": Object {
                            "className": "hljs language-javascript",
                            "data-lang": "javascript",
                          },
                          "t": "code",
                        },
                      ],
                      "p": Object {
                        "className": "highlight",
                      },
                      "t": "pre",
                    },
                    "
",
                  ],
                  "p": Object {
                    "className": "content",
                  },
                  "t": "div",
                },
                "
",
              ],
              "p": Object {
                "className": "listingblock",
              },
              "t": "div",
            },
            "
",
          ],
          "p": Object {
            "className": "sectionbody",
          },
          "t": "div",
        },
        "
",
      ],
      "p": Object {
        "className": "sect1",
      },
      "t": "div",
    },
  ],
  "t": "div",
}
`;
